kube-runtime/src/observer.rs

```diff
@@ -0,0 +1,114 @@
+use backoff::ExponentialBackoff;
+use futures::{future::ready, Stream, StreamExt};
+use kube_client::{
+    api::{Api, ListParams},
+    core::Resource,
+};
+use serde::de::DeserializeOwned;
+use std::fmt::Debug;
+
+use crate::{
+    utils,
+    watcher::{watcher, Error, Event, Result},
+};
+
+
+/// A simple observer around a watcher with error handling and retry backoff
+///
+/// # Error handling
+///
+/// An `Observer` sets a sensible default backoff policy for all watch events and will retry
+/// (with expotential backoff) from transient `watcher` failures until the retry policy is breached.
+///
+/// To configure the retry policy use `Observer::backoff`.
+///
+/// Note that while all errors will be retried, some represent the need for user action to recover:
+///
+/// - 404 `ErrorResponse`(watching invalid / missing api kind/group for `K`)
+/// - 403 `ErrorResponse` (missing list + watch rbac verbs for `K`)
+pub struct Observer<K>
+where
+    K: Clone + Resource + Send + Sync + 'static,
+{
+    // temporary builder params
+    api: Api<K>,
+    listparams: Option<ListParams>,
+    backoff: Option<ExponentialBackoff>,
```

Comment:
Probably want this to be a `Box<dyn Backoff>`? The `Option` also seems a bit excessive here too.

Reply:
I wanted to make this more generic, but it had many issues as noted in https://github.com/kube-rs/kube-rs/pull/703#issuecomment-966633046 . Raised a PR for one of them, but haven't even received a response to it. Library hasn't seen __any__ action for 10 months actually (despite its high use). Might have to fork it to get what we need :(

Reply:
https://github.com/ihrwein/backoff/pull/49

Reply:
Ah, that is nice! Let's hope that gets merged and released. Still no releases for 10 months.

Reply:
Asked for a release explicitly in the repo, we rely on 2 or 3 unreleased prs in there now.